src/App.tsx

```diff
@@ -1,24 +1,125 @@
-import React, { useState } from 'react';
+import React, { useReducer } from 'react';
 import { StyleSheet, Text, View } from 'react-native';
 import AddItem from './components/AddItem';
 import AddPartyMember from './components/AddPartyMember';
 import Tax from './components/Tax';
 import Tip from './components/Tip';
 
+enum Actions {
+  ADD_ITEM = 'add_item',
+  ADD_PARTY_MEMBER = 'add_party_member',
+  SET_TIP = 'set_tip',
+  SET_TAX = 'set_tax',
+}
+
+type AppStateType = {
+  total: number;
+  items: {
+    [name: string]: {
+      price: number;
+      purchaserList: string[];
+    };
+  };
+  party: string[];
+  tip: number;
+  tax: number;
+};
+
+type ActionArgs = {
+  type: Actions;
+
+  // item name or party member name
+  name: string;
+
+  // item details
+  price: number;
+  purchaserList: string[];
+
+  // tax or tip amount
+  amount: number;
+};
+
 export default function App() {
-  const [total, setTotal] = useState(0);
-  const [items, setItems] = useState({});
-  const [party, setParty] = useState([]);
-  const [tip, setTip] = useState(0);
-  const [tax, setTax] = useState(0);
-
-  const addItem = (itemName: string, itemPrice: number, itemPurchaserList: Array<string>) => {
-    setItems({ ...items, [itemName]: [itemPrice, itemPurchaserList] });
-    setTotal(Number((total + itemPrice).toFixed(2)));
+  const initialState: AppStateType = {
+    total: 0,
+    items: {},
+    party: [],
+    tip: 0,
+    tax: 0,
+  };
+
+  const reducer = (state: AppStateType, action: Partial<ActionArgs>) => {
+    switch (action.type) {
+      case Actions.ADD_ITEM: {
+        const updatedItems = {
+          ...state.items,
+          [action.name]: {
+            price: action.price,
+            purchaserList: action.purchaserList,
+          },
+        };
+        const total = Object.values(updatedItems).reduce(
```

Comment:
so unlike the previous code where we just did `setTotal(Number((total + itemPrice).toFixed(2)))`, here we are calculating the total from the existing list of objects.

I'm not sure what the expected user flow is but currently when we update the items object by just updating via the item name, the user could overwrite the details of an existing item name and lead to an inaccurate subtotal since we were just adding before.

Reply:
Nice catch and change!